Type hints and cleanup of Barcode.build() and surrounding code

maresb opened the pull request:
This is a follow-up to #228. I'm trying to not break backwards-compatibility for now, but just to make the code more understandable.

WhyNotHugo commented on barcode/__init__.py:
```diff
@@ -89,15 +108,15 @@ def get(
     code: str,
-    writer: BaseWriter | None = None,
-    output: str | (os.PathLike | (BinaryIO | None)) = None,
+    writer: BaseWriter | None,
```
This parameter was previously optional. With this change, it is no longer optional.

Some tests fail because of this. You can run them with `pytest`.

maresb replied:
I'm changing this back.

I'm not defending it, but for disclosure I think for some reason I wanted `output` to never not be `None`.  This forced me to remove the default value. Since positional args can't come after kwargs, this forced me to remove the default value on `writer`.

WhyNotHugo commented on barcode/writer.py:
```diff
@@ -204,65 +219,60 @@ def packed(self, line: str) -> Generator[tuple[int, float], str, None]:
                 c = 1
 
-    def render(self, code):
+    def render(self, code: list[str]) -> Any:
```
Don't use `Any`; just leave the type hint absent until we can address it.

WhyNotHugo reviewed (changes requested):
Sorry for the slow reply, these were a lot of changes and needed serious attention.

The grand majority of this is fine; I only see one problem (which breaks existing usages and tests).

This also makes `mypy` fail for `barcode/writer.py`, but I can address that myself separately (you're only surfacing an existing issus).

maresb commented on barcode/base.py:
```diff
@@ -34,16 +34,21 @@ class Barcode:
         return f"<{self.__class__.__name__}({self.get_fullcode()!r})>"
 
     def build(self) -> list[str]:
+        """Return a singleton list with a string encoding the barcode as 1s and 0s."""
```
Ah, I'm using math terminology here. Thanks for pointing out the conflicting meaning, I'll change it!

> I think that some barcodes have other values (like G) to encode guard lines.

Oh, ya, I see that now in EAN13, thanks!

maresb commented:
> Sorry for the slow reply, these were a lot of changes and needed serious attention.

Ya, honestly I felt bad for dumping this mess on you. Thanks so much for sifting through it!!! This looks quick, let me see if I can crank this out now...

maresb commented:
I got pytest running. Now I'm looking into:

```python
>       raise RuntimeError(
            f"Character {char} could not be converted in charset {self._charset}."
        )
E       RuntimeError: Character 9 could not be converted in charset C.
```

maresb commented:
Ok, cool, I introduced this exception in 5ff0cd7 :joy:

maresb commented on tests/test_manually.py:
```diff
@@ -62,7 +68,7 @@ def append_img(x, y) -> None:
 
     options = {}
-    for codename, code in TESTCODES:
+    if True:
```
I get rid of this in the subsequent "Dedent" commit.

maresb replied:
That subsequent "Dedent" commit looks really nasty on GitHub because it gets confused between `options` and `opts`.

It's much cleaner when I view it in VS Code:

![image](https://github.com/WhyNotHugo/python-barcode/assets/15216687/d41f8a42-97a5-456e-96c9-7f3f9f1ee151)

maresb commented:
@WhyNotHugo, could you please re-review? I tried to make the individual commits very simple.

One potentially confusing commit is "Parametrize test_generating_barcodes". I preserved the HTML by creating a module-scoped fixture that collects the individual HTML image elements. This way the test cases can pass or fail independently.

The other commit that requires some thought is "Cleanup Barcode.build()". The logic here is that only one invocation of `self._convert` can possibly encode a digit from charset C which results in returning `None`. Thus I split it out this case into a separate function `_convert_or_buffer`. While it's more lines-of-code, I think it's easier to understand being more type-friendly and less like a mysterious finite-state automaton.

maresb commented:
If it's too much then let me know and I'll split it into multiple PRs.

WhyNotHugo commented:
I've triggered a rebase after merging https://github.com/WhyNotHugo/python-barcode/pull/233.

Feel free to rebase locally and force-push if you had any pending changes.

maresb commented:
Thanks @WhyNotHugo!!! I agree with the rebase.

I split off the non-minimal changes into #234, so the only unreviewed commits are the last four, directly addressing your feedback above, and I'd expect those to be super-quick to review.

maresb commented:
@WhyNotHugo, there is no urgency whatsoever on this, but it seems like you reviewed all but the last four small commits, so perhaps this is simple to merge.

WhyNotHugo commented:
I think this is okay to merge. There are commits that make changes and later commits that revert them. Do you want to rebase this? Otherwise I'll just squash them into a single commit.

WhyNotHugo commented:
Thanks!
